optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/uni/UniConstraintStream.java

```diff
@@ -180,7 +190,332 @@
      * @return a stream that matches every combination of A and B for which all the {@link BiJoiner joiners} are true
      */
     default <B> BiConstraintStream<A, B> join(Class<B> otherClass, BiJoiner<A, B>... joiners) {
-        return join(otherClass, AbstractBiJoiner.merge(joiners));
+        int joinerCount = joiners.length;
+        int indexOfFirstFilter = -1;
+        for (int i = 0; i < joinerCount; i++) {
+            BiJoiner<A, B> joiner = joiners[i];
+            boolean hasAFilter = indexOfFirstFilter >= 0;
+            if (joiner instanceof FilteringBiJoiner) {
+                if (!hasAFilter) { // From now on, we only allow filtering joiners.
+                    indexOfFirstFilter = i;
+                }
+            } else if (hasAFilter) {
+                throw new IllegalStateException("Indexing joiner (" + joiner + ") must not follow a filtering joiner ("
+                        + joiners[indexOfFirstFilter] + ").");
+            }
```

Comment:
For readability I'd turn these if's order and avoid the boolean hasAFilter:

```
if (indexOfFirstFilter >= 0) {
      if (joiner instanceof FilteringBiJoiner) {
        throw ...
      }
} else if (joiner instanceof FilteringBiJoiner) {
    indexOfFirstFilter = i;
}
```

Most of OptaPlanner's code has the style to first check the fail fasts, before doing the code. I've found this is more readable, because when reading you can skip the fail-fasts, reducing the lines to grok (fail-fast's are usually easy to recognize based on their "throw" keyword).

Reply:
I partially agree - with the order, not with avoiding the `hasAFilter` variable, which in my opinion rather adds to readability.